In [14]:
import pandas as pd

In [15]:
df = pd.read_csv("../../backend/data/raw/male_players.csv")
df.head()

/var/folders/v6/c29cbybs71971d_cj04qx27m0000gn/T/ipykernel_73426/2624931925.py:1: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../../backend/data/raw/male_players.csv")


,player_id,player_url,fifa_version,fifa_update,update_as_of,short_name,long_name,player_positions,overall,potential,...,ldm,cdm,rdm,rwb,lb,lcb,cb,rcb,rb,gk
0,231747,/player/231747/kylian-mbappe/240002,24.0,2.0,2023-09-22,K. Mbappé,Kylian Mbappé Lottin,"ST, LW",91,94,...,63+3,63+3,63+3,68+3,63+3,54+3,54+3,54+3,63+3,18+3
1,239085,/player/239085/erling-haaland/240002,24.0,2.0,2023-09-22,E. Haaland,Erling Braut Haaland,ST,91,94,...,63+3,63+3,63+3,62+3,60+3,62+3,62+3,62+3,60+3,19+3
2,192985,/player/192985/kevin-de-bruyne/240002,24.0,2.0,2023-09-22,K. De Bruyne,Kevin De Bruyne,"CM, CAM",91,91,...,80+3,80+3,80+3,79+3,75+3,70+3,70+3,70+3,75+3,21+3
3,158023,/player/158023/lionel-messi/240002,24.0,2.0,2023-09-22,L. Messi,Lionel Andrés Messi Cuccittini,"CF, CAM",90,90,...,63+3,63+3,63+3,64+3,59+3,49+3,49+3,49+3,59+3,19+3
4,165153,/player/165153/karim-benzema/240002,24.0,2.0,2023-09-22,K. Benzema,Karim Benzema,"CF, ST",90,90,...,64+3,64+3,64+3,64+3,60+3,55+3,55+3,55+3,60+3,18+3


In [16]:
df.columns.tolist()

['player_id',
 'player_url',
 'fifa_version',
 'fifa_update',
 'update_as_of',
 'short_name',
 'long_name',
 'player_positions',
 'overall',
 'potential',
 'value_eur',
 'wage_eur',
 'age',
 'dob',
 'height_cm',
 'weight_kg',
 'club_team_id',
 'club_name',
 'league_id',
 'league_name',
 'league_level',
 'club_position',
 'club_jersey_number',
 'club_loaned_from',
 'club_joined_date',
 'club_contract_valid_until_year',
 'nationality_id',
 'nationality_name',
 'nation_team_id',
 'nation_position',
 'nation_jersey_number',
 'preferred_foot',
 'weak_foot',
 'skill_moves',
 'international_reputation',
 'work_rate',
 'body_type',
 'real_face',
 'release_clause_eur',
 'player_tags',
 'player_traits',
 'pace',
 'shooting',
 'passing',
 'dribbling',
 'defending',
 'physic',
 'attacking_crossing',
 'attacking_finishing',
 'attacking_heading_accuracy',
 'attacking_short_passing',
 'attacking_volleys',
 'skill_dribbling',
 'skill_curve',
 'skill_fk_accuracy',
 'skill_long_passing',
 'skill_ball_co

In [17]:
## So we basically have 109 feaures lets exclude unnecessary features
## or in other words reduce inputs to 15-20

In [18]:
df.ldm.sample(5)

148311      66
16682     44+2
1156      71+2
1576      54+2
148626      73
Name: ldm, dtype: object

In [19]:
df.shape

(180021, 109)

In [20]:
df.isnull().sum().sort_values(ascending=False).head(20)

nation_jersey_number              169923
nation_team_id                    169923
nation_position                   169923
club_loaned_from                  169298
player_tags                       166103
goalkeeping_speed                 159997
player_traits                      98216
release_clause_eur                 59299
mentality_composure                32888
shooting                           20024
passing                            20024
physic                             20024
defending                          20024
dribbling                          20024
pace                               20024
club_joined_date                   12588
league_level                        2250
value_eur                           2153
club_contract_valid_until_year      1865
club_jersey_number                  1865
dtype: int64

In [21]:
feature_cols = [c for c in df.columns if c.startswith(('attacking_','skill_','movement_','power_','mentality_','defending_'))]
X = df[feature_cols].fillna(0)

In [22]:
X.columns

Index(['skill_moves', 'attacking_crossing', 'attacking_finishing',
       'attacking_heading_accuracy', 'attacking_short_passing',
       'attacking_volleys', 'skill_dribbling', 'skill_curve',
       'skill_fk_accuracy', 'skill_long_passing', 'skill_ball_control',
       'movement_acceleration', 'movement_sprint_speed', 'movement_agility',
       'movement_reactions', 'movement_balance', 'power_shot_power',
       'power_jumping', 'power_stamina', 'power_strength', 'power_long_shots',
       'mentality_aggression', 'mentality_interceptions',
       'mentality_positioning', 'mentality_vision', 'mentality_penalties',
       'mentality_composure', 'defending_marking_awareness',
       'defending_standing_tackle', 'defending_sliding_tackle'],
      dtype='object')

In [41]:
X.shape

(18350, 30)

In [23]:
X.sample(5)

,skill_moves,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,attacking_volleys,skill_dribbling,skill_curve,skill_fk_accuracy,skill_long_passing,...,power_long_shots,mentality_aggression,mentality_interceptions,mentality_positioning,mentality_vision,mentality_penalties,mentality_composure,defending_marking_awareness,defending_standing_tackle,defending_sliding_tackle
108,4,82,80,54,87,79,86,80,76,82,...,81,45,58,81,86,55,84.0,49,45,56
148066,4,69,75,70,78,59,78,62,73,71,...,70,43,39,80,80,78,0.0,50,55,22
166398,4,70,68,47,63,68,75,69,69,45,...,76,39,37,72,62,62,0.0,25,24,24
128344,2,44,33,42,45,29,57,32,30,38,...,33,48,54,42,40,42,43.0,53,55,57
30244,2,25,23,58,54,27,47,21,29,43,...,19,64,61,23,44,32,56.0,64,65,64


In [9]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [10]:
import psutil

In [26]:
psutil.virtual_memory()

svmem(total=17179869184, available=7096172544, percent=58.7, used=8856076288, free=548569088, active=6539083776, inactive=6260015104, wired=2316992512)

In [27]:
# from sklearn.metrics.pairwise import cosine_similarity
# sim = cosine_similarity(X_scaled)
# doesnt run well 
# and anyways this is all vs all 
# we want one vs all

In [28]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

X = df[feature_cols].fillna(0).astype('float32').values  # float32 halves memory too

def get_similar(player_idx, X, top_k=10):
    query = X[player_idx].reshape(1, -1)
    sims = cosine_similarity(query, X)[0]
    top_indices = np.argsort(sims)[::-1]
    top_indices = top_indices[top_indices != player_idx][:top_k]  # exclude itself
    return df.iloc[top_indices][['short_name', 'overall']], sims[top_indices]

In [32]:
df['fifa_version'].unique()

array([24., 23., 22., 21., 20., 19., 18., 17., 16., 15.])

In [29]:
# Someone of his level
target_name = "Haaland"
idx = df[df['short_name'].str.contains(target_name, case=False, na=False)].index[0]
print(df.loc[idx, ['short_name', 'overall', 'club_name']])

similar_players, scores = get_similar(idx, X, top_k=10)
print(similar_players)
print(scores)

short_name         E. Haaland
overall                    91
club_name     Manchester City
Name: 1, dtype: object
           short_name  overall
18370      E. Haaland       88
36905      E. Haaland       88
111822    C. Immobile       82
129957    C. Immobile       80
30475   V. Postolachi       63
65150       G. Karlen       66
56221      E. Haaland       84
7390    V. Postolachi       67
93552     C. Immobile       87
129861        W. Bony       81
[0.9983462  0.99794173 0.9977365  0.9976484  0.99759257 0.99755716
 0.99750865 0.99742144 0.9973629  0.99735814]


# The way Haaland repeats worries me 
## :(
So now we change stuff
short_name         E. Haaland
overall                    91
club_name     Manchester City
Name: 1, dtype: object
           short_name  overall
18370      E. Haaland       88
36905      E. Haaland       88
111822    C. Immobile       82
129957    C. Immobile       80
30475   V. Postolachi       63
65150       G. Karlen       66
56221      E. Haaland       84
7390    V. Postolachi       67
93552     C. Immobile       87
129861        W. Bony       81
[0.9983462  0.99794173 0.9977365  0.9976484  0.99759257 0.99755716
 0.99750865 0.99742144 0.9973629  0.99735814]


In [33]:
df = df[df['fifa_version'] == df['fifa_version'].max()]
df = df.drop_duplicates(subset=['short_name', 'club_name', 'overall'])
df = df.reset_index(drop=True)

In [36]:
len(df), X.shape

(18350, (180021, 30))

In [37]:
X = df[feature_cols].fillna(0).astype('float32').values
len(df), X.shape

(18350, (18350, 30))

In [38]:
target_name = "Haaland"
idx = df[df['short_name'].str.contains(target_name, case=False, na=False)].index[0]
print(df.loc[idx, ['short_name', 'overall', 'club_name']])

similar_players, scores = get_similar(idx, X, top_k=10)
print(similar_players)
print(scores)

short_name         E. Haaland
overall                    91
club_name     Manchester City
Name: 1, dtype: object
          short_name  overall
7390   V. Postolachi       67
2190        A. Čolak       74
2187   J. Pohjanpalo       74
242        C. Wilson       82
75       C. Immobile       85
2710        D. Selke       73
18        V. Osimhen       88
13190    C. Herrmann       62
1776        Vinícius       75
4462    Pedro Mendes       70
[0.99742144 0.99685675 0.9966228  0.99651223 0.9964875  0.9963919
 0.9961938  0.996112   0.99596864 0.9958773 ]


In [40]:
import pickle

with open('../models/player_baseline.pkl', 'wb') as f:
    pickle.dump({
        'X': X,
        'feature_cols': feature_cols,
        'scaler': scaler
    }, f)

df.to_csv('players_clean.csv', index=False)
print("Saved.")

Saved.
